src/coreclr/src/System.Private.CoreLib/src/System/Threading/ProcessorIdCache.cs

```diff
@@ -0,0 +1,124 @@
+        // In relatively rare cases of a slow GetCurrentProcessorNumber, it may be adjusted to a higher number.
+        private static int s_processorIdRefreshRate = 50;
+        // We will not adjust higher than this though.
+        private const int MaxIdRefreshRate = 5000;
+
+        private static int RefreshCurrentProcessorId()
+        {
+            int currentProcessorId = Thread.GetCurrentProcessorNumber();
+
+            // On Unix, GetCurrentProcessorNumber() is implemented in terms of sched_getcpu, which
+            // doesn't exist on all platforms.  On those it doesn't exist on, GetCurrentProcessorNumber()
+            // returns -1.  As a fallback in that case and to spread the threads across the buckets
+            // by default, we use the current managed thread ID as a proxy.
+            if (currentProcessorId < 0) currentProcessorId = Environment.CurrentManagedThreadId;
+
+            Debug.Assert(s_processorIdRefreshRate <= ProcessorIdCacheCountDownMask);
+
+            // Mask with int.MaxValue to ensure the execution Id is not negative
+            t_currentProcessorIdCache = ((currentProcessorId << ProcessorIdCacheShift) & int.MaxValue) | s_processorIdRefreshRate;
+
+            return currentProcessorId;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static int GetCurrentProcessorId()
+        {
+            int currentProcessorIdCache = t_currentProcessorIdCache--;
+            if ((currentProcessorIdCache & ProcessorIdCacheCountDownMask) == 0)
+            {
+                return RefreshCurrentProcessorId();
+            }
+
+            return currentProcessorIdCache >> ProcessorIdCacheShift;
+        }
+
+        // If GetCurrentProcessorNumber takes any nontrivial time (compared to TLS access), return false.
+        // Check more than once - to make sure it was not because TLS was delayed by GC or a context switch.
+        internal static unsafe int ProcessorNumberSpeedCheck()
+        {
+            // NOTE: We do not check the frequency of the Stopwatch.
+            //       The frequency often does not match the actual timer refresh rate anyways.
+            //       If the resolution, precision or access time to the timer are inadequate for our measures here,
+            //       the test will fail anyways.
+
+            // default values pretend to be very long samples with default ratio
+            double minID = Stopwatch.Frequency * 50;  // 50 sec
+            double minTLS = Stopwatch.Frequency;      // 1 sec
+
+            // warm up the code paths.
+            UninlinedThreadStatic();
+            if (Thread.GetCurrentProcessorNumber() < 0)
+                return MaxIdRefreshRate;
+
+            long oneMicrosecond = Stopwatch.Frequency / 1000000;
+            for (int i = 0; i < 10; i++)
+            {
+                // we will measure at least 16 iterations and at least 1 microsecond
```

Comment:
You will actually always measure at least 32 iterations because `iters` get multiplied by 2 as the first thing in the loop.